### PR TITLE
Implement approximate file matching #5

### PR DESCRIPTION
**Pull request by JonoYang**

### CODE REVIEW

File: src/matchcode_toolkit/fingerprinting.py [hunk 1]
```diff
@@ -158,3 +163,75 @@ def create_halohash_chunks(bah128):
...
+    ngs_bytes = [[g.encode('utf-8') for g in ng] for ng in ngs]
+
+    # Join all ngrams into a single bytearray
+    ngs_bytes = [b''.join(ng) for ng in ngs_bytes]
+
+    # Create fingerprints and return fingerprint hashes
+    file_fingerprint = BitAverageHaloHash(ngs_bytes) if ngs_bytes else None
+
+    return dict(
+        halo1=file_fingerprint.hexdigest().decode('utf-8') if file_fingerprint else ''
```
Comment (JonoYang) [inline src/matchcode_toolkit/fingerprinting.py:236]: @pombredanne What should the file fingerprint be called? Currently it is `halo1` but I don't think that's a great label.

**Review by JonoYang — COMMENTED** [review]
@pombredanne I have a question regarding the file fingerprint name

File: src/matchcode_toolkit/fingerprinting.py (hunk 1 above)
Comment (pombredanne) [reply src/matchcode_toolkit/fingerprinting.py:236]: > What should the file fingerprint be called? Currently it is `halo1` but I don't think that's a great label.

This is OK for now IMHO. 

> Also, if a fingerprint could not be generated for a file, should we still populate that resource's extra_data field with `{'halo1': ''}` or just avoid returning anything at all?

IMHO, always provide a value, null or empty (the same we do for other hashes)

**Comment by JonoYang** [comment]
@pombredanne What additional tests do you suggest?

**Comment by pombredanne** [comment]
@JonoYang what about a test that would be using two real code files where the second one has been modified a little?

**Pull request merged**